script/src/types.rs

```diff
@@ -112,17 +112,109 @@ pub(crate) type Machine = AsmMachine;
 #[cfg(not(has_asm))]
 pub(crate) type Machine = TraceMachine<CoreMachine>;
 
+/// Common data that would be shared amongst multiple VM instances.
+/// One sample usage right now, is to capture suspended machines in
+/// a chain of spanwed machines.
+#[derive(Default)]
+pub struct MachineContext {
+    pub(crate) suspended_machines: Vec<ResumableMachine>,
+}
+
+/// Data structure captured all environment data for a suspended machine
+#[derive(Clone, Debug)]
+pub enum ResumePoint {
+    Initial,
+    Spawn {
+        callee_peak_memory: u64,
+        callee_memory_limit: u64,
+        content: Vec<u8>,
+        content_length: u64,
+        caller_exit_code_addr: u64,
+        caller_content_addr: u64,
+        caller_content_length_addr: u64,
+    },
+}
+
+/// Data structure captured all environment data for a suspended machine.
+/// While ResumePoint is used in snapshots for easier serialization, this
+/// contains data that can be used at runtime amongst running machines.
+#[derive(Clone, Debug)]
+pub enum ResumeData {
```

Comment:
I suggest to remove this enum, we can replace it by changing the struct `ResumableMachine` to enum, this will make the code shorter and easier to read:
```rust
pub enum ResumableMachine {
    Intial(Machine),
    Spawn(Machine, SpawnData),
}
```

Reply:
I do suggest we keep this enum, the problem lies at [spawn syscalls](https://github.com/nervosnetwork/ckb/blob/077d20554aaa7df6daf4114fd20c5596657cb2a7/script/src/syscalls/spawn.rs#L246-L255), here we are using `ResumeData` to build the `Machine` that gets included in `ResumableMachine`, moving the data from `ResumeData` into `ResumableMachine` means we will have to pass in a ton of parameters here. Keeping `ResumeData` as its own data structure, IMO, is a more balanced solution.

Reply:
there is a typo in my comment, this is an updated version:
```
pub enum ResumableMachine {
    Intial(Machine),
    Spawn(Machine, ResumeData),
}

pub struct ResumeData {
    callee_peak_memory: u64,
    callee_memory_limit: u64,
    content: Arc<Mutex<Vec<u8>>>,
    content_length: u64,
    caller_exit_code_addr: u64,
    caller_content_addr: u64,
    caller_content_length_addr: u64,
}
```

Reply:
Done